File: src/sdk/ts/src/embedding-artifact.ts
import { readFile, writeFile } from "node:fs/promises";
import { mergeEmbeddingArtifacts, type Skill, type Tool } from "../native/index.cjs";
import type { EmbeddingSpec } from "./catalog.js";
import { mapArtifactError } from "./errors.js";
import { SkillRegistry, ToolRegistry } from "./registry.js";

/**
 * Build-time embedding artifact for catalog warm (ADR-0018). Exactly one of
 * `path` or `bytes`. Default `onMiss` is `"error"`.
 */
export type ExperimentalEmbeddingArtifact =
  | {
      /** Filesystem path to an embedding artifact. */
      readonly path: string;
      /** @ignore */
      readonly bytes?: never;
      /** `"error"` (default) or `"embed"` missing corpus ids only. */
      readonly onMiss?: "error" | "embed";
    }
  | {
      /** In-memory embedding artifact bytes. */
      readonly bytes: Uint8Array;
      /** @ignore */
      readonly path?: never;
      /** `"error"` (default) or `"embed"` missing corpus ids only. */
      readonly onMiss?: "error" | "embed";
    };

/** Options for {@link experimentalBuildEmbeddingArtifact}. */
export interface ExperimentalBuildEmbeddingArtifactOptions {
  /** Output path (`fs.writeFile`; parent must exist). */
  output: string;
  /** Same as catalogs; omit for the built-in default model. */
  embedding?: EmbeddingSpec;
  /** Tool metadata (no `execute`). Empty/omitted is valid. */
  tools?: readonly Tool[];
  /** Skill corpus. Empty/omitted is valid; both empty → empty RAT1. */
  skills?: readonly Skill[];
}

/**
 * Build one mixed Tool+Skill RAT1 and write it to `output`. Builds a Tool half
 * and a Skill half via the registry single-kind
 * {@link ToolRegistry.experimentalBuildEmbeddingArtifact | experimentalBuildEmbeddingArtifact}
 * builders (metadata registered on BM25 registries so each document is embedded
 * once, not via catalog `buildDense`), merges them internally, and writes the
 * combined artifact.
 *
 * @throws {EmbedderError} When embedding or backend resolution fails.
 * @throws {ArtifactError} When artifact construction fails for a non-embedder reason.
 * @throws {IncompatibleMergeError} When the Tool and Skill halves are incompatible.
 */
export async function experimentalBuildEmbeddingArtifact(
  options: ExperimentalBuildEmbeddingArtifactOptions,
): Promise<void> {
  const tools = options.tools ?? [];
  const skills = options.skills ?? [];
  const embedding = options.embedding;

  const toolRegistry = new ToolRegistry(embedding, "bm25");
  if (tools.length > 0) {
    toolRegistry.registerItems(tools);
  }
  const toolBytes = await toolRegistry.experimentalBuildEmbeddingArtifact();

  const skillRegistry = new SkillRegistry(embedding, "bm25");
  if (skills.length > 0) {
    skillRegistry.registerItems(skills);
  }
  const skillBytes = await skillRegistry.experimentalBuildEmbeddingArtifact();

  const merged = (() => {
    try {
      return mergeEmbeddingArtifacts([toolBytes, skillBytes]);
    } catch (error) {
      throw mapArtifactError(error);
    }
  })();
  await writeFile(options.output, merged);
}

function validateEmbeddingArtifact(config: unknown): {
  path?: string;
  bytes?: Uint8Array;
  onMiss: "error" | "embed";
} {
  if (config === null || typeof config !== "object" || Array.isArray(config)) {
    throw new TypeError("experimentalEmbeddingArtifact must be an object");
  }
  const record = config as Record<string, unknown>;
  const unknown = Object.keys(record).filter(
    (key) => key !== "path" && key !== "bytes" && key !== "onMiss",
  );
  if (unknown.length > 0) {
    throw new TypeError(
      `experimentalEmbeddingArtifact has unknown keys: ${unknown.sort().join(", ")}`,
    );
  }
  const path = record.path;
  const bytes = record.bytes;
  const hasPath = path !== undefined;
  const hasBytes = bytes !== undefined;
  if (hasPath === hasBytes) {
    throw new TypeError("experimentalEmbeddingArtifact requires exactly one of 'path' or 'bytes'");
  }
  const onMissRaw = record.onMiss === undefined ? "error" : record.onMiss;
  if (onMissRaw !== "error" && onMissRaw !== "embed") {
    throw new Error(
      `unknown on-artifact-miss policy ${JSON.stringify(onMissRaw)} (expected "error" or "embed")`,
    );
  }
  if (hasPath) {
    if (typeof path !== "string") {
      throw new TypeError("experimentalEmbeddingArtifact path must be a string");
    }
    return { path, onMiss: onMissRaw };
  }
  const raw = bytes;
  if (!(raw instanceof Uint8Array)) {
    throw new TypeError("experimentalEmbeddingArtifact bytes must be a Uint8Array or Buffer");
  }
  return { bytes: raw, onMiss: onMissRaw };
}

export async function resolveEmbeddingArtifact(
  config: ExperimentalEmbeddingArtifact,
): Promise<{ bytes: Buffer; onMiss: "error" | "embed" }> {
  const validated = validateEmbeddingArtifact(config);
  if (validated.path !== undefined) {
    return { bytes: await readFile(validated.path), onMiss: validated.onMiss };
  }
  const bytes = validated.bytes;
  if (bytes === undefined) {
    throw new TypeError("experimentalEmbeddingArtifact requires exactly one of 'path' or 'bytes'");
  }
  return { bytes: Buffer.from(bytes), onMiss: validated.onMiss };
}

File: src/sdk/python/ratel_ai/embedding_artifact.py
"""Build-time embedding artifacts (ADR-0018) — Python experimental surface.

Hosts own filesystem I/O; core remains bytes-only. Catalogs warm a configured
artifact before eager document embedding when
``experimental_embedding_artifact`` is set.
"""

from __future__ import annotations

import asyncio
import os
from collections.abc import Iterable, Mapping
from pathlib import Path
from typing import TYPE_CHECKING, Literal, TypedDict, Union, cast

from ._native import merge_embedding_artifacts

if TYPE_CHECKING:
    from .catalog import EmbeddingSpec, Tool
    from .skill_catalog import Skill

OnArtifactMiss = Literal["error", "embed"]


class _ExperimentalEmbeddingArtifactOptions(TypedDict, total=False):
    on_miss: OnArtifactMiss


class ExperimentalEmbeddingArtifactPath(_ExperimentalEmbeddingArtifactOptions):
    """Filesystem path to a RAT1 embedding artifact."""

    path: str


class ExperimentalEmbeddingArtifactBytes(_ExperimentalEmbeddingArtifactOptions):
    """In-memory RAT1 embedding artifact bytes."""

    bytes: bytes


ExperimentalEmbeddingArtifact = Union[
    ExperimentalEmbeddingArtifactPath,
    ExperimentalEmbeddingArtifactBytes,
]


async def experimental_build_embedding_artifact(
    *,
    output: str | os.PathLike[str],
    embedding: EmbeddingSpec | None = None,
    tools: Iterable[Tool] | None = None,
    skills: Iterable[Skill] | None = None,
) -> None:
    """Build one mixed Tool+Skill RAT1 and write it to ``output``.

    Registers metadata on BM25 registries so each document is embedded at most
    once (not via catalog dense build). Empty/omitted corpora are valid; both
    empty yields a valid empty RAT1. Always builds both registry sides (empty →
    empty RAT1) so merge always receives two parts. Parent directories must
    already exist; no atomic write.

    Raises:
        EmbedderError: Embedding/model/inference failure from either registry build.
        ArtifactError: Non-embedder artifact construction failure from either registry build.
        IncompatibleMergeError: Valid Tool/Skill artifact parts cannot be internally merged.
        OSError: Filesystem failure writing ``output`` (not wrapped as artifact errors).
    """
    from .catalog import ToolRegistry
    from .skill_catalog import SkillRegistry

    tool_items = list(tools or ())
    skill_items = list(skills or ())

    tool_registry = ToolRegistry(embedding, method="bm25")
    if tool_items:
        await tool_registry.register(tool_items)
    tool_bytes = await tool_registry.experimental_build_embedding_artifact()

    skill_registry = SkillRegistry(embedding, method="bm25")
    if skill_items:
        await skill_registry.register(skill_items)
    skill_bytes = await skill_registry.experimental_build_embedding_artifact()

    merged = merge_embedding_artifacts([tool_bytes, skill_bytes])
    await asyncio.to_thread(Path(output).write_bytes, merged)


def _validate_embedding_artifact(
    config: ExperimentalEmbeddingArtifact,
) -> tuple[str | None, bytes | None, OnArtifactMiss]:
    if not isinstance(config, Mapping):
        raise TypeError("experimental_embedding_artifact must be a mapping")
    unknown = set(config) - {"path", "bytes", "on_miss"}
    if unknown:
        raise TypeError(
            f"experimental_embedding_artifact has unknown keys: {sorted(unknown)}"
        )
    path_raw = config.get("path")
    bytes_raw = config.get("bytes")
    has_path = path_raw is not None
    has_bytes = bytes_raw is not None
    if has_path == has_bytes:
        raise TypeError(
            "experimental_embedding_artifact requires exactly one of 'path' or 'bytes'"
        )
    on_miss_raw = config.get("on_miss")
    if on_miss_raw is None:
        on_miss_raw = "error"
    if on_miss_raw not in ("error", "embed"):
        raise ValueError(
            f'unknown on-artifact-miss policy {on_miss_raw!r} (expected "error" or "embed")'
        )
    assert on_miss_raw == "error" or on_miss_raw == "embed"
    on_miss: OnArtifactMiss = on_miss_raw
    if has_path:
        path = cast(ExperimentalEmbeddingArtifactPath, config)["path"]
        if not isinstance(path, str):
            raise TypeError("experimental_embedding_artifact path must be a str")
        return path, None, on_miss
    raw = cast(ExperimentalEmbeddingArtifactBytes, config)["bytes"]
    if not isinstance(raw, bytes):
        raise TypeError("experimental_embedding_artifact bytes must be bytes")
    return None, raw, on_miss


async def resolve_embedding_artifact(
    config: ExperimentalEmbeddingArtifact,
) -> tuple[bytes, OnArtifactMiss]:
    """Load artifact bytes and resolve ``on_miss`` (default ``"error"``)."""
    path, raw, on_miss = _validate_embedding_artifact(config)
    if raw is not None:
        return raw, on_miss
    assert path is not None
    return await asyncio.to_thread(Path(path).read_bytes), on_miss
